server/base/src/main/java/org/apache/accumulo/server/ServiceEnvironmentImpl.java

```diff
@@ -0,0 +1,140 @@
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.StreamSupport;
+
+import org.apache.accumulo.core.client.TableNotFoundException;
+import org.apache.accumulo.core.clientImpl.Table;
+import org.apache.accumulo.core.clientImpl.Tables;
+import org.apache.accumulo.core.conf.AccumuloConfiguration;
+import org.apache.accumulo.core.conf.ConfigurationTypeHelper;
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.core.spi.common.ServiceEnvironment;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMap.Builder;
+
+public class ServiceEnvironmentImpl implements ServiceEnvironment {
+
+  final ServerContext srvCtx;
+
+  public ServiceEnvironmentImpl(ServerContext ctx) {
+    this.srvCtx = ctx;
+  }
+
+  private static class ConfigurationImpl implements Configuration {
+
+    private final AccumuloConfiguration acfg;
+    private Map<String,String> customProps;
+    private Map<String,String> tableCustomProps;
+
+    ConfigurationImpl(AccumuloConfiguration acfg) {
+      this.acfg = acfg;
+    }
+
+    @Override
+    public String get(String key) {
+      // Get prop to check if sensitive, also looking up by prop may be more efficient.
+      Property prop = Property.getPropertyByKey(key);
+      if (prop != null) {
+        if (prop.isSensitive())
+          return null;
+        return acfg.get(prop);
+      } else {
+        return acfg.get(key);
+      }
+    }
+
+    @Override
+    public Map<String,String> getCustom() {
+      if (customProps == null)
+        customProps = buildCustom(Property.GENERAL_ARBITRARY_PROP_PREFIX);
+
+      return customProps;
+    }
+
+    @Override
+    public String getCustom(String keySuffix) {
+      return getCustom().get(keySuffix);
+    }
+
+    @Override
+    public Map<String,String> getTableCustom() {
+      if (tableCustomProps == null)
+        tableCustomProps = buildCustom(Property.TABLE_ARBITRARY_PROP_PREFIX);
+
+      return tableCustomProps;
+    }
+
+    @Override
+    public String getTableCustom(String keySuffix) {
+      return getTableCustom().get(keySuffix);
+    }
+
+    private Map<String,String> buildCustom(Property customPrefix) {
+      // This could be optimized as described in #947
+      Map<String,String> props = acfg.getAllPropertiesWithPrefix(customPrefix);
+      Builder<String,String> builder = ImmutableMap.builder();
+      props.forEach((k, v) -> {
+        builder.put(k.substring(customPrefix.getKey().length()), v);
+      });
+
+      return builder.build();
+    }
+
+    @Override
+    public Iterator<Entry<String,String>> iterator() {
+      return StreamSupport.stream(acfg.spliterator(), false)
+          .filter(e -> !Property.isSensitive(e.getKey())).iterator();
+    }
+  }
+
+  @Override
+  public Configuration getConfiguration() {
+    return new ConfigurationImpl(srvCtx.getConfiguration());
```

Comment:
Do we want to create a new object when getConfiguration() methods are called?  I am thinking of the use case where user code passes around the ServiceEnv and calls getConfiguration().get() frequently would create too many objects.  Shouldn't we create a single ConfigurationImpl object in the ServiceEnvironment constructor?

Reply:
Changed in 6c08827